### PR TITLE
Ability to retrieve assets information through API (closes #566)

### LINKED ISSUE #566: Ability to retrieve solvency information through API

### PR DESCRIPTION
**Pull request by ysv**

### CODE REVIEW

File: app/api/api_v2/assets.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
+module APIv2
```
Comment (yivo) [inline app/api/api_v2/assets.rb]: I think we should go with peatio convention for it:

There should not any index proofs actions. Only get.

Get latest proof should be: `/solvency/liability_proofs/latest?currency=btc`

You must not allow fiat currencies to be accepted. Please, don't waste time for writing detailed description. Set desc for params to: `The code of any currency with type set to "coin".`. This param is mandatory.

Get my partial tree: `/solvency/liability_proofs/partial_tree/mine?currency=btc`.

Please, update the names for grape entity and API classes (we need to keep all solvency entities namespaced).

**Review by yivo — CHANGES_REQUESTED** [review]

File: app/api/api_v2/assets.rb (hunk 1 above)
Comment (yivo) [reply app/api/api_v2/assets.rb]: Also please write some very simple specs.

**Comment by ysv** [comment]
@yivo review please

File: app/api/api_v2/solvency.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+module APIv2
+  class Solvency < Grape::API
+
+    before { authenticate! }
+
+    params do
+      requires :currency, type: String, values: -> { Currency.coin_codes(bothcase: true) }, desc: "The code of any currency with type 'coin'."
+    end
+    get '/solvency/liability_proofs/latest' do
```
Comment (yivo) [inline app/api/api_v2/solvency.rb]: @yssavchuk It should return only the latest proofs, one per currency, check `Proof.current`

File: app/api/api_v2/solvency.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+      Proof
+        .order(id: :desc)
+        .find_by_currency_id(Currency.find_by!(code: params[:currency]).id)
+        .tap { |p| present p, with: APIv2::Entities::Solvency::Proof }
+    end
+
+    params do
+      requires :currency, type: String, values: -> { Currency.coin_codes(bothcase: true) }, desc: "The code of any currency with type 'coin'."
+    end
+    get '/solvency/liability_proofs/partial_tree/mine' do
```
Comment (yivo) [inline app/api/api_v2/solvency.rb]: @yssavchuk It should return single partial tree extracted from the latest proof for specified currency. Check `Private::AssetsController#partial_tree`.

File: app/api/api_v2/solvency.rb (hunk 2 above)
Comment (yivo) [reply app/api/api_v2/solvency.rb]: Check Check `Private::AssetsController#index`.

**Review by yivo — CHANGES_REQUESTED** [review]

**Comment by yivo** [comment]
@yssavchuk Also please fix this:
image

File: app/api/api_v2/solvency.rb (hunk 3 above)
Comment (ysv) [reply app/api/api_v2/solvency.rb]: ```Private::AssetsController#partial_tree``` is broken
deepinscreenshot_select-area_20180322135335

Comment (ysv) [reply app/api/api_v2/solvency.rb]: I will submit issue

File: app/api/api_v2/solvency.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      current_user
+        .accounts
+        .with_currency(params[:currency])
```
Comment (yivo) [inline app/api/api_v2/solvency.rb]: @yssavchuk Better to use `Member#get_account` here.

File: spec/api/api_v2/solvency_spec.rb [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+      it 'require currency code to be present and to be coin' do
+        api_get '/api/v2/solvency/liability_proofs/latest', token: token
+        expect(response.code).to eq '422'
+
+        api_get '/api/v2/solvency/liability_proofs/latest', params: { currency: 'usd' }, token: token
+        expect(response.code).to eq '422'
+      end
+
+      it 'should not fails if proofs doesn\'t exist' do
```
Comment (yivo) [inline spec/api/api_v2/solvency_spec.rb]: `should not fail if proofs don't exist`

Comment (yivo) [reply spec/api/api_v2/solvency_spec.rb]: `doesn\'t fail if proofs don\'t exist`

File: spec/api/api_v2/solvency_spec.rb [hunk 6]
```diff
@@ -0,0 +1,83 @@
+describe APIv2::Solvency, type: :request do
+
+  let!(:member) { create(:member, :verified_identity) }
+  let(:token) { jwt_for(member) }
+
+  describe 'GET api/v2/solvency/liability_proofs/latest' do
+    context 'without liability proof' do
```
Comment (yivo) [inline spec/api/api_v2/solvency_spec.rb]: `when no liability proofs available in database`

File: spec/api/api_v2/solvency_spec.rb [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+      it 'should not fails if proofs doesn\'t exist' do
+        api_get '/api/v2/solvency/liability_proofs/latest', params: { currency: 'btc' }, token: token
+
+        expect(response.code).to eq '200'
+        proof = JSON.parse(response.body)
+        expect(proof).to eq nil
+      end
+    end
+
+    context 'with liability proof' do
```
Comment (yivo) [inline spec/api/api_v2/solvency_spec.rb]: `when liability proofs available in database`

File: spec/api/api_v2/solvency_spec.rb [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+      let!(:proofs) { create_list(:proof, 10) }
+
+      it 'should return liability proof and it should be freshest' do
```
Comment (yivo) [inline spec/api/api_v2/solvency_spec.rb]: `latest`

File: spec/api/api_v2/solvency_spec.rb [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        expect(response.code).to eq '200'
+        proof = JSON.parse(response.body)
+        expect(proof).not_to eq nil
+        expect(proof['id']).to eq proofs.last.id
+      end
+    end
+  end
+
+  describe 'GET api/v2/solvency/liability_proofs/partial_tree/mine' do
+    context 'without partial tree' do
```
Comment (yivo) [inline spec/api/api_v2/solvency_spec.rb]: `no partial trees available in database`

File: spec/api/api_v2/solvency_spec.rb [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+      it 'should not fails if partial trees doesn\'t exist' do
+        api_get '/api/v2/solvency/liability_proofs/partial_tree/mine', params: { currency: 'btc' }, token: token
+
+        expect(response.code).to eq '200'
+        partial_tree = JSON.parse(response.body)
+        expect(partial_tree).to eq nil
+      end
+    end
+
+    context 'with partial tree' do
```
Comment (yivo) [inline spec/api/api_v2/solvency_spec.rb]: `when partial trees available in database`

File: spec/api/api_v2/solvency_spec.rb [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+      let!(:partial_trees) { create_list(:partial_tree, 10, account: member.accounts.find_by_currency_id(Currency.find_by_code(:btc).id)) }
```
Comment (yivo) [inline spec/api/api_v2/solvency_spec.rb]: `member.ac(:btc).id`

File: spec/api/api_v2/solvency_spec.rb [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+      it 'should return partial trees and it should be freshest' do
```
Comment (yivo) [inline spec/api/api_v2/solvency_spec.rb]: `latest`

File: app/api/api_v2/solvency.rb [hunk 13]
```diff
@@ -0,0 +1,28 @@
+module APIv2
+  class Solvency < Grape::API
+
+    before { authenticate! }
+
+    params do
+      requires :currency, type: String, values: -> { Currency.coin_codes(bothcase: true) }, desc: "The code of any currency with type 'coin'."
+    end
+    get '/solvency/liability_proofs/latest' do
```
Comment (yivo) [inline app/api/api_v2/solvency.rb]: @yssavchuk Add desc: `Returns newest liability proof record for given currency.`.

File: app/api/api_v2/solvency.rb [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+      Proof
+        .current(params[:currency])
+        .tap { |p| present p, with: APIv2::Entities::Solvency::Proof }
+    end
+
+    params do
+      requires :currency, type: String, values: -> { Currency.coin_codes(bothcase: true) }, desc: "The code of any currency with type 'coin'."
+    end
+    get '/solvency/liability_proofs/partial_tree/mine' do
```
Comment (yivo) [inline app/api/api_v2/solvency.rb]: @yssavchuk Desc needed: `Returns newest partial tree record for member account of specified currency.`.

**Review by yivo — CHANGES_REQUESTED** [review]

**Review by yivo — APPROVED** [review]

**Comment by yivo** [comment]
@yssavchuk I rebased the branch.

**Comment by yivo** [comment]
@yssavchuk Seems OK.

image

**Comment by yivo** [comment]
image

**Pull request merged**